AdapterRemoval v2
=================

.. image:: https://img.shields.io/travis/MikkelSchubert/adapterremoval/master.svg
    :target: https://travis-ci.org/MikkelSchubert/adapterremoval
    :alt: Travis-CI

.. image:: https://img.shields.io/coveralls/MikkelSchubert/adapterremoval.svg
    :target: https://coveralls.io/github/MikkelSchubert/adapterremoval
    :alt: Coveralls

AdapterRemoval searches for and removes adapter sequences from High-Throughput
Sequencing (HTS) data and (optionally) trims low quality bases from the 3' end
of reads following adapter removal. AdapterRemoval can analyze both single end
and paired end data, and can be used to merge overlapping paired-ended reads
into (longer) consensus sequences. Additionally, AdapterRemoval can construct a
consensus adapter sequence for paired-ended reads, if which this information is
not available.

If you use AdapterRemoval v2, then please cite the paper

    Schubert, Lindgreen, and Orlando (2016). AdapterRemoval v2: rapid adapter
    trimming, identification, and read merging. BMC Research Notes, 12;9(1):88
    http://bmcresnotes.biomedcentral.com/articles/10.1186/s13104-016-1900-2

AdapterRemoval was originally published in Lindgreen 2012:

    Lindgreen (2012): AdapterRemoval: Easy Cleaning of Next Generation
    Sequencing Reads, BMC Research Notes, 5:337
    http://www.biomedcentral.com/1756-0500/5/337/


Overview of major features
==========================

 - Trimming of adapters sequences from single-end and paired-end FASTQ reads.
 - Trimming of multiple, different adapters or adapter pairs.
 - Demultiplexing of single or double indexed reads, with or without trimming
   of adapter sequences.
 - Reconstruction of adapter sequences from paired-end reads, by the pairwise
   alignment of reads in the absence of a known adapter sequence.
 - Merging of overlapping read-pairs into higher-quality consensus sequences.
 - Multi-threading of all operations for increased throughput.
 - Reading and writing of gzip and bzip2 compressed files.
 - Reading and writing of interleaved FASTQ files.


Installation
============

For detailed installation instructions, please see the
`Installation <https://adapterremoval.readthedocs.io/en/latest/installation.html>`_
section of the online documentation.

Installing AdapterRemoval using `Conda <https://conda.io/docs/>`_
-----------------------------------------------------------------

If you have Conda `installed on your system <https://conda.io/miniconda.html>`_,
you may install AdapterRemoval using one of serveral unoffical recipies::

    conda install -c bioconda adapterremoval
    conda install -c maxibor adapterremoval2


Installing AdapterRemoval on Debian based systems
-------------------------------------------------

Users of Debian Stretch or Ubuntu Artful Aardvark may install AdapterRemoval
using Apt::

    sudo apt-get install adapterremoval adapterremoval-examples


Building AdapterRemoval from scratch
------------------------------------

Compiling AdapterRemoval from scratch requires a C++11 compliant compiler, and
that the zlib and bz2lib headers are installed::

    make
    sudo make install


Getting started
===============

To run AdapterRemoval, specify the location of pair 1 and (optionally) pair 2
FASTQ using the --file1 and --file2 command-line options::

    AdapterRemoval --file1 myreads_1.fastq.gz --file2 myreads_2.fastq.gz

By default, AdapterRemoval will save the trimmed reads in the current working
directly, using filenames starting with 'your_output'. See the
`Input and Output
<https://adapterremoval.readthedocs.io/en/latest/input_and_output.html>`_
files section for more information about files generated by AdapterRemoval.

More examples of common usage may be found in the
`Examples <https://adapterremoval.readthedocs.io/en/latest/examples.html>`_
section of the online documentation.


Documentation
-------------

For a detailed description of program usage, please refer to the online
`online documentation <https://adapterremoval.readthedocs.io/>`_.
